src/seq.rs

```diff
@@ -9,61 +9,337 @@
+pub trait IteratorExt: Iterator + Sized {
+    /// Choose one element at random from the iterator.
+    ///
+    /// Returns `None` if and only if the iterator is empty.
+    /// 
+    /// Complexity is `O(n)`, where `n` is the length of the iterator.
+    /// This likely consumes multiple random numbers, but the exact number
+    /// is unspecified.
+    fn choose<R>(mut self, rng: &mut R) -> Option<Self::Item>
+        where R: Rng + ?Sized
+    {
+        if let Some(elem) = self.next() {
+            let mut result = elem;
+            
+            // Continue until the iterator is exhausted
+            for (i, elem) in self.enumerate() {
+                let k = rng.gen_range(0, i + 2);
+                if k == 0 {
+                    result = elem;
+                }
+            }
+            
+            Some(result)
+        } else {
+            None
+        }
+    }
+
+    /// Collects `amount` values at random from the iterator into a supplied
+    /// buffer.
+    /// 
+    /// Although the elements are selected randomly, the order of elements in
+    /// the buffer is neither stable nor fully random. If random ordering is
+    /// desired, shuffle the result.
+    /// 
+    /// Returns the number of elements added to the buffer. This equals `amount`
+    /// unless the iterator contains insufficient elements, in which case this
+    /// equals the number of elements available.
+    /// 
+    /// Complexity is `O(n)` where `n` is the length of the iterator.
+    fn choose_multiple_fill<R>(mut self, rng: &mut R, buf: &mut [Self::Item])
+        -> usize where R: Rng + ?Sized
+    {
+        let amount = buf.len();
+        let mut len = 0;
+        while len < amount {
+            if let Some(elem) = self.next() {
+                buf[len] = elem;
+                len += 1;
+            } else {
+                // Iterator exhausted; stop early
+                return len;
+            }
+        }
+
+        // Continue, since the iterator was not exhausted
+        for (i, elem) in self.enumerate() {
+            let k = rng.gen_range(0, i + 1 + amount);
+            if k < amount {
+                buf[k] = elem;
+            }
+        }
+        len
+    }
+
+    /// Collects `amount` values at random from the iterator into a vector.
+    ///
+    /// This is equivalent to `choose_multiple_fill` except for the result type.
+    ///
+    /// Although the elements are selected randomly, the order of elements in
+    /// the buffer is neither stable nor fully random. If random ordering is
+    /// desired, shuffle the result.
+    /// 
+    /// The length of the returned vector equals `amount` unless the iterator
+    /// contains insufficient elements, in which case it equals the number of
+    /// elements available.
+    /// 
+    /// Complexity is `O(n)` where `n` is the length of the iterator.
+    #[cfg(feature = "alloc")]
+    fn choose_multiple<R>(mut self, rng: &mut R, amount: usize) -> Vec<Self::Item>
```

Comment:
I tried combining the implementations too. Problem is we want allocated-on-first-use storage which `Vec` gives us for free but isn't compatible with buffers; some kind of trait might be able to abstract over this, but as you said it just results in more code in total.

About the return type, I'm not sure; both have their merits. However what about consistency? I think it would be possible to make `SliceExt::choose_multiple` return a `Result` too, even using iterators as it does now.

Reply:
Good point about consistency. Would make sense to return a Result<Iterator, Iterator> for `SliceExt::choose_multiple`.

You could even argue that we should be consistent with `choose` and return an `Result<T, Something>`, but that's more dubious.

I don't have strong feelings about this. Maybe sticking with what we have now makes for the cleanest API.

Reply:
The `std` library already uses `Option` in cases where `None` is kind-of an error. I don't think it makes sense to use `Result<T, ()>` (or `NoInput` instead of `()`).

Reply:
Yeah, I was thinking that too. I think just returning a Vec/Iterator might be the cleanest solution.